internal/printer/print-to-js.go

```diff
@@ -131,45 +130,89 @@ func render1(p *printer, n *Node, opts RenderOptions) {
 
 		for c := n.FirstChild; c != nil; c = c.NextSibling {
 			if c.Type == TextNode {
-				p.addNilSourceMapping()
 				p.printInternalImports(p.opts.InternalURL, &opts)
 
+				start := 0
 				if len(n.Loc) > 0 {
-					p.addSourceMapping(n.Loc[0])
+					start = c.Loc[0].Start
 				}
 				render := js_scanner.HoistImports([]byte(c.Data))
-				importStatements := ""
 				if len(render.Hoisted) > 0 {
-					for _, hoisted := range render.Hoisted {
-						statement := string(bytes.TrimSpace(hoisted)) + "\n"
-						importStatements += statement
+					for i, hoisted := range render.Hoisted {
+						if len(bytes.TrimSpace(hoisted)) == 0 {
+							continue
+						}
+						hoistedLoc := render.HoistedLocs[i]
+						p.printTextWithSourcemap(string(hoisted)+"\n", loc.Loc{Start: start + hoistedLoc.Start})
 					}
 				}
-				preprocessed := js_scanner.HoistExports(render.Body)
-
-				if len(c.Loc) > 0 {
-					p.addSourceMapping(c.Loc[0])
-				}
-				p.println(strings.TrimSpace(importStatements))
 
 				if opts.opts.StaticExtraction {
+					p.addNilSourceMapping()
 					p.printCSSImports(opts.cssLen)
 				}
 
 				// 1. Component imports, if any exist.
-				p.printComponentMetadata(n.Parent, opts.opts, []byte(importStatements))
+				p.addNilSourceMapping()
+				p.printComponentMetadata(n.Parent, opts.opts, []byte(p.sourcetext))
 				// 2. Top-level Astro global.
+
 				p.printTopLevelAstro(opts.opts)
 
-				if len(preprocessed.Hoisted) > 0 {
-					for _, hoisted := range preprocessed.Hoisted {
-						p.println(strings.TrimSpace(string(hoisted)))
+				exports := make([][]byte, 0)
+				exportLocs := make([]loc.Loc, 0)
+				bodies := make([][]byte, 0)
+				bodiesLocs := make([]loc.Loc, 0)
+
+				if len(render.Body) > 0 {
+					for i, innerBody := range render.Body {
+						innerStart := render.BodyLocs[i].Start
+						if len(bytes.TrimSpace(innerBody)) == 0 {
+							continue
+						}
+
+						// Extract exports
+						preprocessed := js_scanner.HoistExports(append(innerBody, '\n'))
+						if len(preprocessed.Hoisted) > 0 {
+							for j, exported := range preprocessed.Hoisted {
+								exportedLoc := preprocessed.HoistedLocs[j]
+								exportLocs = append(exportLocs, loc.Loc{Start: start + innerStart + exportedLoc.Start})
+								exports = append(exports, exported)
+							}
+						}
+
+						if len(preprocessed.Body) > 0 {
+							for j, body := range preprocessed.Body {
+								bodyLoc := preprocessed.BodyLocs[j]
+								bodiesLocs = append(bodiesLocs, loc.Loc{Start: start + innerStart + bodyLoc.Start})
+								bodies = append(bodies, body)
+							}
+						}
```

Comment:
This is interleaving import hoisting with export hoisting, which gets kinda hard to follow, but the main point is that we're preserving the location of each statement.